solver: use penalties for variant values

alalazo opened the pull request:
fixes #51112

On develop nodes in a DAG get a penalty of 1 when a variant value is not the default. This might lead to cases of non-determinism, due to having multiple optimal solutions.

Here we give different penalties to different non-default values, thereby reducing the non-determinism in concretization.

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

alalazo committed:
solver: use penalties for variant values

fixes #51112

On develop nodes in a DAG get a penalty of 1 when a variant
value is not the default. This might lead to cases of
non-determinism, due to having multiple optimal solutions.

Here we give different penalties to different non-default values,
thereby reducing the non-determinism in concretization.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo commented:
This bugfix comes with a slight regression:

* **Spack:** 1.2.0.dev0 (https://github.com/spack/spack/commit/e62ca3b64cc516926180a9cc0bf5b7ebad5d22ae)
* **Builtin repo:** https://github.com/spack/spack-packages/commit/127fe0c1bf32638f9c092d49dd9a04b5b1404f81
* **Python:** 3.12.3
* **Platform:** linux-ubuntu24.04-broadwell

[radiuss.develop.csv](https://github.com/user-attachments/files/24230725/radiuss.develop.csv)
[radiuss.pr.csv](https://github.com/user-attachments/files/24230726/radiuss.pr.csv)

<img width="2000" height="1000" alt="comparison" src="https://github.com/user-attachments/assets/c390fe97-31ad-4d14-9911-7725d8f28d09" />

haampie commented on lib/spack/spack/solver/asp.py:
```diff
@@ -1704,11 +1704,17 @@ def define_variant(
+            pkg_fact(fn.variant_default_value_from_package_py(vid, val))
+
+        if variant_def.values is not None:
+            current_values = sorted(variant_def.values, key=lambda v: (v not in default_values, v))
```
Why sorted? Using the provided order makes more sense to me.

alalazo replied:
Lower penalties to values marked as defaults in `package.py`

haampie replied:
That's clear, but then for non-defaults you disrupt the order of possible values defined in `package.py`. E.g. `values=("c99", "c11", "c17")` or `values=("20", "17", "14", "11")` etc.

alalazo replied:
Do we want to rely on that order? I.e. do we want:
```
values=("11", "14", "17")
```
to concretize differently from:
```
values=("14", "11", "17")
```
? Not sure if that's better or just another "weird" decision to explain to users. I think we shouldn't rely on the order of declared values for weights.

haampie replied:
It gives control to package maintainers. What's hard to explain is why given `variant("c_std", values=("99, "11", "17")` Spack would consider `11`, then `17`, then `99`.

alalazo replied:
I don't really agree, but I also don't care much as long as we have a deterministic order for single valued variants, when the default values cannot be used. Pushed a change to just put default values first.

haampie replied:
Okay good. Then maybe do it in linear time instead and non-allocating instead of n lg n? Looping twice is probably faster.

alalazo replied:
For reference:
```python
import timeit
from spack.solver.core import fn

default_values = ("17", )
values = ("11", "14", "17", "20", "23")

def with_sorted():
    current_values = sorted(values, key=lambda v: v not in default_values)
    for penalty, v in enumerate(current_values, 1):
        fn.pkg_fact(fn.variant_penalty(v, penalty))

def with_double_loop():
    penalty = 1
    for v in default_values:
        fn.pkg_fact(fn.variant_penalty(v, penalty))
        penalty += 1

    for v in values:
        if v not in default_values:
            fn.pkg_fact(fn.variant_penalty(v, penalty))
            penalty += 1
```
we have:
```
$ spack python -m timeit -s "from script import with_sorted" "with_sorted()"
20000 loops, best of 5: 11.6 usec per loop
$ spack python -m timeit -s "from script import with_double_loop" "with_double_loop()"
20000 loops, best of 5: 10.7 usec per loop
```
Checking on a specific spec (`trilinos`), the impact on speed-up doesn't seem significant on `setup`. Is the increment in lines of code worth it? If you think so, I'll just push the change but I'd rather leave it like it is for now.

haampie replied:
I think it is. I'm getting the following on Python 3.14:

> With sorted: 0.847418 seconds
> With double loop: 0.735292 seconds

using

```python
if __name__ == "__main__":
    sorted_time = timeit.timeit(with_sorted, number=100000)
    double_loop_time = timeit.timeit(with_double_loop, number=100000)

    print(f"With sorted: {sorted_time:.6f} seconds")
    print(f"With double loop: {double_loop_time:.6f} seconds")
```

It also becomes more clear that there's an order, whereas with `sorted` you have to remember that Python's sort is a stable sort.

haampie replied:
It seems though that `Variant` internally sorts the values, and `DisjointSetsOfValues` uses sets an effectively randomizes the values.

```
>>> spack.repo.PATH.get_pkg_class("libfabric").variants[spack.spec.Spec("@:")]["fabrics"].values
('cxi', 'efa', 'gni', 'lnx', 'mlx', 'mrail', 'opx', 'psm', 'psm2', 'psm3', 'rxm', 'rxd', 'shm', 'sockets', 'tcp', 'ucx', 'udp', 'usnic', 'verbs', 'xpmem')
```

```
>>> spack.repo.PATH.get_pkg_class("sirius").variants[spack.spec.Spec("+cuda")]["cuda_arch"].values.sets
[{'none'}, {'53', '100', '101', '110', '120a', '120f', '32', '120', '37', '21', '60', '87', '20', '12', '52', '61', '89', '100f', '101a', '35', '90', '110a', '100a', '86', '121f', '90a', '121', '75', '121a', '103a', '50', '80', '103f', '13', '72', '62', '11', '70', '30', '103', '101f', '110f', '10'}]
```

alalazo committed:
Use the order declared in directive

This requires modifications to DinjointSetsOfValues to
store the values in the order they are declared.

A few unit tests have been modified to ensure we check
the order is preserved.

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented:
Last comment here before Christmas break:

I would still try to avoid the `sorted` call on the values. Yes, sorted leads to determinism, but since Spack has no clue what the values represent, it is effectively the same as randomization: if the default fabric in libfabric does not work, you get the next one in alphabetical order. If  the C standard `C17` does not work, you get `C99` because it comes next in alphabetical order, and if that doesn't, you get `C11`. That makes no sense.

However, apparently we do not keep track of the order as defined in package.py but throw values into sets, meaning that the scope of this PR becomes larger.

I think that's worth it, since merging this with `sorted` and fixing it later means there's an unnecessary concretizer change.

For what it's worth, if someone else wants to review this PR in the meantime and has different opinions, that's fine. I'll catch up early January.

alalazo committed:
Micro-optimize variant penalty loop

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
solver: simplify variant logic

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

haampie commented on lib/spack/spack/test/concretization/core.py:
```diff
@@ -314,6 +315,16 @@ def gcc11_with_flags(compiler_factory):
 
+def weights_from_result(result: Result, *, name: str) -> Dict[str, int]:
+    weights = {}
+    for x in [x for x in result.criteria if x.name == name]:
```
nitpick: the intermediate list looks a bit odd to me

haampie commented on lib/spack/spack/variant.py:
```diff
@@ -508,23 +512,28 @@ class DisjointSetsOfValues(collections.abc.Sequence):
-            raise spack.error.SpecError("sets in input must be disjoint")
+        cumulated: Set[str] = set()
+        for current_set in self.sets:
+            intersection = cumulated.intersection(set(current_set))
```
I think `set` is redundant here.

haampie replied:
Maybe this?

```python
for curr in self.sets:
  if not cumulated.isdisjoint(curr):
    duplicates = ", ".join(sorted(cumulated.intersection(curr)))
```

haampie commented on lib/spack/spack/variant.py:
```diff
@@ -582,7 +591,7 @@ def __getitem__(self, idx):
         return tuple(itertools.chain.from_iterable(self.sets))[idx]
 
     def __len__(self):
-        return len(itertools.chain.from_iterable(self.sets))
```
This raised in the past, right? So it was never called? Makes me wonder whether it's possible to remove the `collections.abc.Sequence` base class? 🤔

alalazo replied:
`len` was not called before, but it is now. I think before we were just using the "iterable" behavior. I guess we can refactor the class in a following PR?

alalazo committed:
Address review comments

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>

alalazo committed:
Fixup

Signed-off-by: Massimiliano Culpo <massimiliano.culpo@gmail.com>
